flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1586,6 +1610,8 @@ public boolean updateState(TaskExecutionState state) {
 
 		if (attempt != null) {
 			try {
+				maybeReleasePartitions(attempt, state);
```

Comment:
this must be done after the attempt is marked as finished. If the finishing fails we might still need these partitions.